### DISCUSSION TITLE
reload CAs only?

### DISCUSSION THREAD

**Discussion by jlearman** [strongswan/strongswan#1805] [Q&A] [answered]
I'm using golang govici to control Strongswan, though I'm willing to use `swanctl` for some operations if it makes sense.

I'd like to be able to delete individual CAs and end certificates, but that's clearly not possible.  Given that, I'd like to be able to clear (or reload) all CAs at one point in time, and all end certificates later.

I thought that I could use `swanctl --load-authorities` and it would clear old CAs and load any CAs it finds in `/etc/strongswan/swanctl/x509ca`, but it doesn't.  It does clear them all, but it doesn't read any files in x509ca.

`swanctl --load-creds` does, but that would also delete all certificates.

So, what does `--load-authorities` actually load, if not CAs?  Since it clears them, why doesn't it load them?

To make matters worse, once `--load-creds` loads a CAs, nothing seems to clear it.  Deleting the file and issuing either `--load-creds` or `--load-certs` doesn't clear it.  Nor does `--load-all`.

My /etc/strongswan/swanctl/conf.d directory is empty except for one empty file.

Any suggestions?

**Comment by tobiasbrunner** [comment, marked as answer]
> I'm using golang govici to control Strongswan, though I'm willing to use `swanctl` for some operations if it makes sense.

Shouldn't make a difference as both use vici (i.e. you could replicate what swanctl does via govici). But mixing them might not be a good idea (except for stuff like the `--list-*` commands) as swanctl won't know anything about what you configured directly via vici.

> I'd like to be able to delete individual CAs and end certificates, but that's clearly not possible.

You can unload individual CAs if you load them via `load-authority` and not just via `load-cert` command. Then the `unload-authority`  command will remove the CA certificate.

> I thought that I could use `swanctl --load-authorities` and it would clear old CAs and load any CAs it finds in `/etc/strongswan/swanctl/x509ca`, but it doesn't. It does clear them all, but it doesn't read any files in x509ca.

`--load-authorities` (re-)loads `authority` sections in swanctl.conf. That's unrelated to that directory. Certificates there are loaded by `--load-creds` (which uses the `load-cert` command to load the certificates it finds in the `x509*` subdirs). Note that `--load-authorities` won't be able to unload CA certificates stored in the `x509ca` directory if you ever use `--load-creds`.

> swanctl --load-creds does, but that would also delete all certificates.

Only if you pass `--clear` (see below). Without it, it won't remove any certificates.

> To make matters worse, once `--load-creds` loads a CAs, nothing seems to clear it.

Certificates loaded via `load-cert` can only be cleared by `clear-creds`, which clears all credentials loaded by the `load-cert`, `load-key`, `load-token`, and `load-shared` commands (as the name indicates). It's what `--load-creds` does before loading credentials anew if `--clear` is passed.

**Comment by jlearman** [comment]
Thanks, understood about swanctl vs vici.  The documentation makes that clear enough.  It might not cover all cases in detail, but there are good enough warnings in the relevant places.

`unload-authority` Perfect!  Why didn't I see that?  Thanks!

If `clear-creds` doesn't clear authorities created using `load-authority`, then I'm all set.  It's fine that certs and PSKs are all cleared, because we manage all those at the same time, independently of the authorities.

Thanks!
